### ISSUE TITLE
Provide type hints

### ISSUE THREAD

**Issue by Kixunil** [petertodd/python-bitcoinlib#243]
I hope the benefits for possibly security-critical library are obvious.

**Comment by mcelrath** [comment]
Agreed. A pull request would be welcome ;-)

**Comment by ysangkok** [comment]
I have submitted PR #247 which could be seen as a preliminary thing to do before doing type hints. Full type hinting needs Python 3.6, and type hinting on just functions needs Python 3 anyway, so Python 2 support may as well be removed since it won't even parse on there if this is added in a nice way.

**Comment by ysangkok** [comment]
One could rebase https://github.com/Simplexum/python-bitcointx/commit/b275373c62ed57ce617f20e4963fb47975cf46cc

**Comment by dgpv** [comment]
>  so Python 2 support may as well be removed

"v0.10.2

Note: this will be the last release of python-bitcoinlib with Python 2.7 compatibility."

I presume that since v0.11.0 python-bitcoinlib can be considered python3-only

**Comment by dgpv** [comment]
> One could rebase Simplexum@b275373

I don't think that it would be that simple, there was significant incompatibe changes made at this stage.

But it could cover some portion of the code, and maybe that could save some code-editing time.

It would need careful review afterwards, as any automated, but not-straightforward code change

**Comment by dgpv** [comment]
Python2 is declared unsupported since v0.11.0 (in release-notes for v0.10.2, "Note: this will be the last release of python-bitcoinlib with Python 2.7 compatibility."), I believe what @ysangkok meant is that the code that deals with python2/3 compatibility shall be removed, and this is the focus of #247 

There's a style of type hints that can be specified as comments, like `a = None # type: Optional[int]`, and were used before python3.6 that allowed type hints in the normal code.

Python3.6 is also the lowest supported version of python3 (https://pythoninsider.blogspot.com/2020/10/python-35-is-no-longer-supported.html), so I think that targeting it, and ditching version 3.5 and lower is sensible.

**Comment by dgpv** [comment]
Because bitcointx uses type hints directly expressed in source (only supported since python3.6) taking type hint code from bitcointx would transfer the python3.6 requirement of bitcointx to bitcoinlib.

**Comment by Kixunil** [comment]
Sounds good, even Debian stable, which is quite conservative has Python 3.7.3 already, so it should be safe, I think.
